src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs

```diff
@@ -2012,28 +2019,26 @@ private static string CheckForMatches(uint[] runningHash, int upTo)
             /// <returns>The string matching the hash, or null.</returns>
             public static string Match(string text)
             {
-                // The longest pattern is 29 characters.
                 // The values in the array are the computed hashes of length
                 // index-1 (so runningHash[0] holds the hash for length 1).
-                var runningHash = new uint[29];
+                Span<uint> runningHash = stackalloc uint[MAX_LEN];
 
                 int longestPossiblePattern = 0;
                 for (int i = 0; i < text.Length; i++)
                 {
+                    // Notice, '-' == ('-' | 0x20)
                     uint h = text[i];
-                    if (h >= 'A' && h <= 'Z')
-                    {
-                        h |= 0x20; // ToLower
-                    }
-                    else if (!((h >= 'a' && h <= 'z') || h == '-'))
+                    h |= 0x20;
+                    if (!((uint)(h - 'a') <= 'z' - 'a' || h == '-'))
                     {
                         // If the character isn't in any of our patterns,
                         // don't bother hashing and reset the running length.
                         longestPossiblePattern = 0;
                         continue;
                     }
 
-                    for (int j = Math.Min(i, runningHash.Length) - 1; j > 0; j--)
+                    Span<uint> rh = runningHash.Slice(0, Math.Min(i, runningHash.Length));
+                    for (int j = rh.Length - 1; j > 0; j--)
```

Comment:
According to https://github.com/PowerShell/PowerShell/pull/18665/files#r1035381609, the `- 1` part should be removed.

```suggestion
                    for (int j = rh.Length; j > 0; j--)
```

Reply:
Do you mean:
```suggestion
                    for (int j = rh.Length - 1; j >= 0; j--)
```

Reply:
`rh[j]` will throw.

Reply:
Yeah, it will throw with your changes. With the original code, remove the `-1` part will fix the issue.